### PR TITLE
refactor the following layout specs to use 'expect_offense':

### PR DESCRIPTION
**Pull request by mechos3d**
Part of #8127
Refactored specs in spec/rubocop/cop/layout/ to use 'expect_offense' matcher :  

- end_of_line_spec.rb, 
- extra_spacing_spec.rb, 
- indentation_width_spec.rb,
- trailing_empty_lines_spec.rb, 
- trailing_whitespace_spec.rb
-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [ ] Added tests.
* [ ] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: spec/rubocop/cop/layout/trailing_empty_lines_spec.rb [hunk 1]
```diff
@@ -106,25 +99,33 @@
     end
 
     it 'registers an offense for multiple trailing blank lines' do
-      offenses = inspect_source(<<~RUBY)
+      expect_offense(
+        ['x = 0',
```
Comment (marcandre) [inline spec/rubocop/cop/layout/trailing_empty_lines_spec.rb:103]: Please use heredoc instead of array

File: spec/rubocop/cop/layout/end_of_line_spec.rb [hunk 2]
```diff
@@ -1,56 +1,63 @@
 # frozen_string_literal: true
 
 RSpec.describe RuboCop::Cop::Layout::EndOfLine, :config do
+  let(:message)          { 'Carriage return character missing.' }
```
Comment (marcandre) [inline spec/rubocop/cop/layout/end_of_line_spec.rb:4]: Could you please copy paste these messages instead? Most specs are like this. Note you can abbreviate with `[...]` when the message is long.

File: lib/rubocop/rspec/expect_offense.rb [hunk 3]
```diff
@@ -113,6 +113,8 @@ def format_offense(source, **replacements)
 
       # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
       def expect_offense(source, file = nil, severity: nil, **replacements)
+        source = "#{source.join("\n")}\n" if source.is_a?(Array)
```
Comment (marcandre) [inline lib/rubocop/rspec/expect_offense.rb:116]: I'm :-1: on this, I much prefer if we use Heredocs instead of arrays of lines which aren't nearly as readable. Let me know if I missed a use-case.

File: spec/rubocop/cop/layout/end_of_line_spec.rb [hunk 4]
```diff
@@ -1,56 +1,63 @@
...
-      expect(cop.offenses.empty?).to be(true)
+      expect_no_offenses('')
     end
   end
 
   shared_examples 'iso-8859-15' do |eol|
     it 'can inspect non-UTF-8 encoded source with proper encoding comment' do
-      inspect_source_file(["# coding: ISO-8859-15#{eol}",
-                           "# Euro symbol: \xa4#{eol}"].join("\n"))
+      inspect_source_file(<<~RUBY)
```
Comment (marcandre) [inline spec/rubocop/cop/layout/end_of_line_spec.rb:15]: Would it be possible to change this to use `expect_offense`?

**Review by marcandre — CHANGES_REQUESTED** [review]
Good PR. There are a few things I'm requesting, we can discuss them, or else please apply them everywhere. Thanks!

File: lib/rubocop/rspec/expect_offense.rb (hunk 3 above)
Comment (mechos3d) [reply lib/rubocop/rspec/expect_offense.rb:116]: Now I agree that it looks more like a hacky workaround than a good solution (because it is :) ).          
As to why I did this: for trailing whitespaces. 
In cases like this - it's not clearly seen are there whitespaces, how much of them are there.       
Also some folks have removing trailing whitespace done automatically by the IDE on save - so one must be careful with these specs.
```ruby
    expect_offense(<<~RUBY)
        
      ^^ Trailing whitespace detected.\n
    RUBY
```
But later I've found a helper method `trailing_whitespace` in (`/spec/support/misc_helper.rb`) - I think it will solve the problem like this: 
 ```ruby
    expect_offense(<<~RUBY)
      #{trailing_whitespace}  
      ^^ Trailing whitespace detected.\n
    RUBY
```    
What do you think ?

File: spec/rubocop/cop/layout/end_of_line_spec.rb (hunk 4 above)
Comment (mechos3d) [reply spec/rubocop/cop/layout/end_of_line_spec.rb:15]: Yeah, it would be.   
I wanted to do it in a separate PR, but we can solve this one right here actually.         
The main goal of this spec - is to test how rubocop works with a file in non-utf8 encoding.  
So reading from a file and respecting magic comment seems like an important part of the process.      

Although maybe it's safe to test it this way: 
```ruby
expect_offense(<<~RUBY.force_encoding('ISO-8859-15'))
   .... 
RUBY
```       
Dealing with the magic comments is not part of rubocop's functionality, so there's no need to test this, we can rely on ruby itself on that. 
What do you think, is the `force_encoding` approach ok?

Comment (mechos3d) [reply spec/rubocop/cop/layout/end_of_line_spec.rb:15]: `def expect_offense(source, file = nil, severity: nil, **replacements)`
Oh, I see now.     
Another way would be just to send some temporary file path to `expect_offense` as an argument..

Comment (marcandre) [reply spec/rubocop/cop/layout/end_of_line_spec.rb:15]: Oh, I missed the encoding part. No problem then (maybe add a comment)

File: lib/rubocop/rspec/expect_offense.rb (hunk 3 above)
Comment (marcandre) [reply lib/rubocop/rspec/expect_offense.rb:116]: Indeed. BTW RuboCop will flag and auto-correct trailing whitespaces in heredocs now too.

File: spec/rubocop/cop/layout/trailing_empty_lines_spec.rb [hunk 5]
```diff
@@ -30,67 +30,60 @@
...
 
       RUBY
-      expect(offenses.size).to eq(1)
-      expect(offenses.first.message).to eq('3 trailing blank lines detected.')
+
+      expect_correction("\n")
     end
 
     it 'registers an offense for no final newline after assignment' do
       offenses = inspect_source('x = 0')
```
Comment (mechos3d) [inline spec/rubocop/cop/layout/trailing_empty_lines_spec.rb:57]: There's another interesting place - the case doesn't work with `expect_offense`, because typing
```
expect_offense(<<~RUBY)
  x = 0
       ^ Final newline is missing
RUBY
```
Would actually have a newline after `x = 0` and will not cause an offense.  

I have some thoughts on how to work around it (possibly with another specific matcher), but I would like to leave it to a separate PR

Comment (marcandre) [reply spec/rubocop/cop/layout/trailing_empty_lines_spec.rb:57]: 👍 😆

Comment (marcandre) [reply spec/rubocop/cop/layout/trailing_empty_lines_spec.rb:57]: That's perfect then.

**Comment by marcandre** [comment]
Great work, thank you!
